### ISSUE TITLE
Scale down the usage of checked exceptions as much as possible

### ISSUE THREAD

**Issue by PawelLipski** [VirtusLab/git-machete-intellij-plugin#1317]
## Steps
- [ ] Remove all (?) `throws` clauses from methods
- [ ] Review all `try...catch` stanzas (most should probably be removed)
- [x] Review all `Try...` calls (most can probably be removed)
- [ ] Where exceptions that we don't control are thrown (like `IOException`), use `@SneakyThrows`
- [ ] For exceptions that we control, either change their base to `RuntimeException` or use `@SneakyThrows`, **TBD**
- [ ] Document the guideline in CONTRIBUTING.md
- [ ] Optionally, enforce some rules with ArchUnit or Checkstyle

## Pros
- [ ] Kotlin/Scala work well without checked exceptions
- [ ] Unclutter the codebase from `Try`s and `try`s

## Cons
- [ ] Some error handling might be gone (like messages of `GitMacheteExceptions` that wrap a `GitCoreException`); probably we can sacrifice that
- [ ] Esp. see places like `com.virtuslab.gitmachete.frontend.file.highlighting.MacheteAnnotator#getParentBranchName`, here we rely on knowing that a checked exception is thrown so that we can deliberately ignore that error 🤔 

WDYT @mkondratek ❓

**Comment by mkondratek** [comment]
wow, that's an interesting idea 😮 at first I thought you ve gone mad XD 

I have been taught (at the university!) to use and throw checked exceptions. And I think we did very nice exception handling and hierarchy (back in the times, with Michał P) - by the book, clear, transparent.

Kotlin/Scala-point seems to be valid, but I do not feel fully informed to consciously agree. I mean - there must be some way, probably some mechanism that is not present in Java that allows working with exceptions in a different way. Or, at least some reasoning that justifies why there are no such exceptions in these languages. Give me some time to learn more about it (I will start from [[here]](https://kotlinlang.org/docs/exceptions.html#checked-exceptions). 

I do agree that try/vavr.Try usage got a bit messy. BUT, I think we could get rid of vavr in that scope. But, bear in mind that it's a different issue and the combination of these changes with the checked exceptions thing would create a huge PR. Even though we decided to proceed with the drop of vavr.Try in the favor of a usual try, and abandon the checked exceptions, we should introduce the changes on separate PRs. 

---

tl;dr
1. Let me get more info (and reasoning) on checked exception pros/cons (in the scope of Java/Kotlin/Scala), or pls share it if you are aware of them.
2. For now I can agree on vavr.Try drop (but with no harm to the checked exceptions).

**Comment by PawelLipski** [comment]
Vavr `Try` is sometimes useful (in the contexts where `Try.sequence` is subsequently applied)... but can be eliminated wherever it's just a replacement for `try ... catch` indeed

**Comment by PawelLipski** [comment]
Nope, IMO the recent #1826 issue provides a good showcase on why checked exceptions aren't such a bad idea, after all 😅
